### ISSUE TITLE
.Net HdrHistogram with multiple writers 

### ISSUE THREAD

**Issue by etishor** [HdrHistogram/HdrHistogram#58]
First of all thanks @giltene and @mattwarren for the awesome work on HdrHistogram. 

I'm trying to implement support for HdrHistogram into my Metrics.Net port of the java metrics library. Initial version uses a lock to sincronize writes and reads to the histogram and so far seems to work fine and is already faster than the default reservoir (exponentially decaying reservoir port). 

I've noticed that the java version uses a "recorder" to handle multiple writer threads and "flipping" the active and inactive histograms. Is there anything similar that can be used in .Net? If not, do you think it is worth porting the java recorder to .net?

Also i wonder, are there any other suggestions on how to handle the scenario where multiple threads record values and one (or maybe more) threads need to read the values?

Thanks,
Iulian

**Comment by mattwarren** [comment]
Yeah the .Net port is a bit behind the Java version, so some of the newer
features are missing. I haven't had any spare time to give to it recently,
but my plan is to make them compatible.

If you'd like to contribute ping me an email (see my github profile) and we
can sync up. The first thing I need to do is move the C# code to a separate
repo, at the moment it's in with the Java code

Previously I found that it was possible to port the Java code to C#
line-by-line, so hopefully that'll be the case for the new stuff Gil came
up with

**Comment by etishor** [comment]
Hi Matt,

I'm interested in having a .net port of the HdrHistogram. It is a very useful tool by itself and I also want to have HdrHistogram in the Metrics.NET library. While integrating in the metrics library I found an issue with getValueAtPercentile() and  had to re-port the java code (https://github.com/etishor/Metrics.NET/commit/5f37bcc69652aa8a475a0df7cbe48c241ad0516f ).

Porting line-by-line has the advantage of keeping the code diff-able with the java version, but results in a harder to change, debug or understand code. It is neither java nor c# :)

What I did for the metrics library was isolate the "complicated" stuff like math, statistics etc, port it line-by-line but write the simple stuff from scratch in a more .net/c# idiomatic way. After I had a good understanding of the complex stuff I ended up re-writing/re-factoring most of it anyway. 

At the moment I'm still reading the java code for the hdr histogram and I think a similar pattern can be applied. 

I'm not sure how much time I'll have in the near future for this, but i'll get in touch over email if I'll have some time to spare.
